Source/Dafny/Translator.cs

```diff
@@ -17354,7 +17354,7 @@ class UsesHeapVisitor : BottomUpVisitor
         if (call != null && call.Function != null && call.Function.ReadsHeap) {
           foundHeap = true;
         }
-        if (expr is ApplyExpr) {
+        if (expr is ApplyExpr || expr is SeqConstructionExpr) {
```

Comment:
Are we sure this is all the expressions where we need to set `foundHeap`?  Could this be a property of the `Expr` subtypes instead so there's some semantic content?  I'm OK with the change, but it seems ad hoc.

Reply:
Each `Expr` has many different characteristics like this. They could be represented in the code as properties of `Expr`, as computed by visitors, or as "type case" statements (that is, cascading `if` statements that perform `is` type tests. Which is done which way is a bit ad hoc. I think there are too many such characteristics to fit comfortably into properties of `Expr` and a number of them also depend on context, so they would need to take parameters whose types don't seem to belong in the AST. So, I suggest we continue with the ad-hoc style, except maybe in cases where it's clear that we could improve the organization.

That said, if there's _duplication_ somewhere, then we should work on eliminating that.